File: PriorityQueues/PriorityQueues.go
package main

import (
	"fmt"
	"time"
	"unsafe"
)

// Demonstrates priority queue operations: insert and extract_min.
// Inserts values and extracts them in ascending order.
type PriorityQueue struct {
	heap []int
}

func (pq *PriorityQueue) insert(val int) {
	pq.heap = append(pq.heap, val)
	pq.heapifyUp(len(pq.heap) - 1)
}

func (pq *PriorityQueue) extractMin() int {
	if len(pq.heap) == 0 {
		panic("extract from empty priority queue")
	}
	minVal := pq.heap[0]
	lastVal := pq.heap[len(pq.heap)-1]
	pq.heap = pq.heap[:len(pq.heap)-1]
	if len(pq.heap) > 0 {
		pq.heap[0] = lastVal
		pq.heapifyDown(0)
	}
	return minVal
}

func (pq *PriorityQueue) peek() int {
	if len(pq.heap) == 0 {
		panic("peek from empty priority queue")
	}
	return pq.heap[0]
}

func (pq *PriorityQueue) isEmpty() bool {
	return len(pq.heap) == 0
}

func (pq *PriorityQueue) heapifyUp(index int) {
	for index > 0 {
		parent := (index - 1) / 2
		if pq.heap[index] < pq.heap[parent] {
			pq.heap[index], pq.heap[parent] = pq.heap[parent], pq.heap[index]
			index = parent
		} else {
			break
		}
	}
}

func (pq *PriorityQueue) heapifyDown(index int) {
	size := len(pq.heap)
	for {
		left := 2*index + 1
		right := 2*index + 2
		smallest := index
		if left < size && pq.heap[left] < pq.heap[smallest] {
			smallest = left
		}
		if right < size && pq.heap[right] < pq.heap[smallest] {
			smallest = right
		}
		if smallest != index {
			pq.heap[index], pq.heap[smallest] = pq.heap[smallest], pq.heap[index]
			index = smallest
		} else {
			break
		}
	}
}

func main() {
	startTime := time.Now()

	pq := &PriorityQueue{}
	values := []int{3, 1, 4, 1, 5, 9, 2, 6}

	for _, val := range values {
		pq.insert(val)
	}

	for !pq.isEmpty() {
		fmt.Println(pq.extractMin())
	}

	elapsed := time.Since(startTime)

	// Memory calculation
	memoryUsed := unsafe.Sizeof(*pq) + uintptr(cap(pq.heap))*unsafe.Sizeof(0)

	// Performance statistics
	fmt.Println("\n--- Performance Statistics ---")
	fmt.Printf("Execution time: %.4f ms\n", float64(elapsed.Nanoseconds())/1e6)
	fmt.Printf("Memory usage: %d bytes\n", memoryUsed)
	fmt.Println("\nTime Complexity:")
	fmt.Println("  - Insert: O(log n)")
	fmt.Println("  - Extract_min: O(log n)")
	fmt.Println("  - Peek: O(1)")
}

File: PriorityQueues/PriorityQueues.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <iomanip>

/**
 * Demonstrates priority queue operations: insert and extract_min.
 * Inserts values and extracts them in ascending order.
 */
class PriorityQueue {
private:
    std::vector<int> heap;

    void heapifyUp(int index) {
        while (index > 0) {
            int parent = (index - 1) / 2;
            if (heap[index] < heap[parent]) {
                std::swap(heap[index], heap[parent]);
                index = parent;
            } else {
                break;
            }
        }
    }

    void heapifyDown(int index) {
        int size = heap.size();
        while (true) {
            int left = 2 * index + 1;
            int right = 2 * index + 2;
            int smallest = index;
            if (left < size && heap[left] < heap[smallest]) {
                smallest = left;
            }
            if (right < size && heap[right] < heap[smallest]) {
                smallest = right;
            }
            if (smallest != index) {
                std::swap(heap[index], heap[smallest]);
                index = smallest;
            } else {
                break;
            }
        }
    }

public:
    void insert(int val) {
        heap.push_back(val);
        heapifyUp(heap.size() - 1);
    }

    int extractMin() {
        if (heap.empty()) {
            throw std::runtime_error("extract from empty priority queue");
        }
        int minVal = heap[0];
        int lastVal = heap.back();
        heap.pop_back();
        if (!heap.empty()) {
            heap[0] = lastVal;
            heapifyDown(0);
        }
        return minVal;
    }

    int peek() {
        if (heap.empty()) {
            throw std::runtime_error("peek from empty priority queue");
        }
        return heap[0];
    }

    bool isEmpty() {
        return heap.empty();
    }
};

int main() {
    auto startTime = std::chrono::high_resolution_clock::now();

    PriorityQueue pq;
    std::vector<int> values = {3, 1, 4, 1, 5, 9, 2, 6};

    for (int val : values) {
        pq.insert(val);
    }

    while (!pq.isEmpty()) {
        std::cout << pq.extractMin() << std::endl;
    }

    auto endTime = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(endTime - startTime);

    // Memory calculation
    size_t memoryUsed = sizeof(PriorityQueue);

    // Performance statistics
    std::cout << "\n--- Performance Statistics ---" << std::endl;
    std::cout << std::fixed << std::setprecision(4);
    std::cout << "Execution time: " << duration.count() / 1000.0 << " ms" << std::endl;
    std::cout << "Memory usage: " << memoryUsed << " bytes" << std::endl;
    std::cout << "\nTime Complexity:" << std::endl;
    std::cout << "  - Insert: O(log n)" << std::endl;
    std::cout << "  - Extract_min: O(log n)" << std::endl;
    std::cout << "  - Peek: O(1)" << std::endl;

    return 0;
}
